### PR TITLE
Making solved  and cardinality Attributes (without tx-layers)

### PR DESCRIPTION
**Pull request by Lojjs**

### CODE REVIEW

File: enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/BuildSlottedExecutionPlan.scala
Comment (sherfert) [inline enterprise/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/BuildSlottedExecutionPlan.scala:89]: We don't need an Attribute to keep track of read only plans. We just need to know if the whole query is read-only.

File: community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/UnionPlanningIntegrationTest.scala
Comment (sherfert) [inline community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/UnionPlanningIntegrationTest.scala]: This does not need to be implicit, does it?

File: community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/idp/CartesianProductsOrValueJoinsTest.scala
Comment (sherfert) [inline community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/idp/CartesianProductsOrValueJoinsTest.scala:133]: This looks like it would work with Fake Attributes.

File: community/cypher/cypher-logical-plans-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/logical/plans/Expand.scala
Comment (fickludd) [inline community/cypher/cypher-logical-plans-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/logical/plans/Expand.scala]: I would suggest not putting all the arguments on one line here, because it becomes much harder to read. Auto-format mishap? Also in more places in this file.

File: community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/idp/IDPQueryGraphSolverTest.scala
Comment (sherfert) [inline community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/idp/IDPQueryGraphSolverTest.scala:47]: This whole file looks like it might work with Fake Attributes.

File: community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/steps/SortSkipAndLimitTest.scala
Comment (sherfert) [inline community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/steps/SortSkipAndLimitTest.scala]: implicit?

File: community/cypher/cypher-planner-3.4/src/main/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/steps/LogicalPlanProducer.scala
Comment (fickludd) [inline community/cypher/cypher-planner-3.4/src/main/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/steps/LogicalPlanProducer.scala:397]: In this class we are keeping `cardinalities`, but that also means we assuming that no earlier attributes exists or that those are no longer needed. Is that the case and/or could we do something to encode that assumption in code somewhere?

File: community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/attribution/Attribute.scala
Comment (fickludd) [inline community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/attribution/Attribute.scala:123]: Yes please, some javadocs around here were needed (and we could add even more later).

File: community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/AstRewritingTestSupport.scala
Comment (fickludd) [inline community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/AstRewritingTestSupport.scala]: I think this guy can be removed without much loss =)

File: community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/AstRewritingTestSupport.scala
Comment (fickludd) [inline community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/AstRewritingTestSupport.scala]: These kinds of of test classes are usually prefixed with `Stub...` rather than `Fake...`. I think this could be changed because it makes naming more consistent.

File: community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/steps/countStorePlannerTest.scala
Comment (fickludd) [inline community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/steps/countStorePlannerTest.scala]: This rename seem like a mistake, and should be reverted?

File: community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/steps/countStorePlannerTest.scala
Comment (fickludd) [inline community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/steps/countStorePlannerTest.scala]: Same here, guess it happened at the same time?

**Review by fickludd — APPROVED** [review]
This PR is a great cleanup in my mind, and reduces the test boilerplate a lot. Let's merge this when green, I don't see anything which has to be fixed before merge.

File: community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/idp/CartesianProductsOrValueJoinsTest.scala
Comment (Lojjs) [reply community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/idp/CartesianProductsOrValueJoinsTest.scala:133]: Nope

File: community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/idp/IDPQueryGraphSolverTest.scala
Comment (Lojjs) [reply community/cypher/cypher-planner-3.4/src/test/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/idp/IDPQueryGraphSolverTest.scala:47]: Also nope, sadly.

File: community/cypher/cypher-planner-3.4/src/main/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/steps/LogicalPlanProducer.scala
Comment (Lojjs) [reply community/cypher/cypher-planner-3.4/src/main/scala/org/neo4j/cypher/internal/compiler/v3_4/planner/logical/steps/LogicalPlanProducer.scala:397]: That is the assumption. An alternative would be to pass in an `Attributes` to that method, but we can do that if if turns out we have other early attributes the LogicalPlanProducer does not know about.

**Pull request merged**
